### PR TITLE
Bumblebees - Esther Kang, Aigerim Kalygulova

### PR DESCRIPTION
**Pull request by essieeekang**

### CODE REVIEW

File: tests/test_wave_01.py [hunk 1]
```diff
@@ -99,7 +99,19 @@ def test_adds_movie_to_non_empty_user_watched():
...
+    user_data = {
+        "watched": [FANTASY_2]
+    }
+
+    # Act
+    updated_data = add_to_watched(user_data, FANTASY_2)
+
+    # Assert
+    assert len(updated_data["watched"]) == 1
+
```
Comment (apradoada) [inline tests/test_wave_01.py:113]: This is a great added test! While we try to cover a variety of potential cases in the tests we have written, there are absolutely times that we just can't cover every edge case. You are always welcome to write more tests that more accurately represent what your code is doing! 

That being said (And this is just for info purposes), the tests that get run against your code are just the ones provided, so do make sure any added tests don't break the ones that are already there!

File: tests/test_wave_01.py [hunk 2]
```diff
@@ -152,19 +176,24 @@ def test_moves_movie_from_watchlist_to_empty_watched():
...
     # Assert
     assert len(updated_data["watchlist"]) == 0
     assert len(updated_data["watched"]) == 1
     
-    raise Exception("Test needs to be completed.")
+    assert updated_data["watched"][0] == {
+            "title": MOVIE_TITLE_1,
+            "genre": GENRE_1,
+            "rating": RATING_1
+        }
```
Comment (apradoada) [inline tests/test_wave_01.py:191]: What I love about this particular assert statement is how specific it is. We know exactly where in the watched list the movie is going to end up, so instead of just checking to see if it's in the list, you've gone ahead and checked the specific index to see if it is what you are expecting. One small tweak you could make here would be to hold the expected movie in its own variable which would allow for a simpler assert statement. I say you "could" make it because at the end of the day, the approach you use will likely be dictated by the team's best practices.

File: tests/test_wave_01.py [hunk 3]
```diff
@@ -183,12 +212,12 @@ def test_moves_movie_from_watchlist_to_watched():
     assert len(updated_data["watchlist"]) == 1
     assert len(updated_data["watched"]) == 2
     
-    raise Exception("Test needs to be completed.")
+    assert updated_data["watched"][-1] == movie_to_watch
```
Comment (apradoada) [inline tests/test_wave_01.py:215]: Looks good to me!

File: tests/test_wave_01.py [hunk 4]
```diff
@@ -205,3 +234,17 @@ def test_does_nothing_if_movie_not_in_watchlist():
...
+        "watchlist": [],
+        "watched": [FANTASY_2]
+    }
+
+    # Act
+    updated_data = watch_movie(janes_data, "Movie")
+
+    # Assert
+    assert len(updated_data["watchlist"]) == 0
+    assert len(updated_data["watched"]) == 1
```
Comment (apradoada) [inline tests/test_wave_01.py:250]: I like this added test, but it does look quite similar to the test right before it! Just be careful to make sure you're not duplicating tests! I know that the main difference between the tests is that the `watchlist` is empty in one. This function should do the same thing whether or not `watchlist` is empty, so I don't think the extra test is necessary in this case!

File: tests/test_wave_02.py [hunk 5]
```diff
@@ -63,3 +63,16 @@ def test_genre_is_None_if_empty_watched():
...
+    tied_data_1 = {"watched": [FANTASY_1, FANTASY_2, ACTION_1, ACTION_2, FANTASY_3, ACTION_3]}
+    tied_data_2 = {"watched": [FANTASY_1, FANTASY_2, ACTION_1, ACTION_2, ACTION_3, FANTASY_3]}
+
+    # Act
+    popular_genre_1 = get_most_watched_genre(tied_data_1)
+    popular_genre_2 = get_most_watched_genre(tied_data_2)
+
+    # Assert
+    assert popular_genre_1 == "Fantasy"
+    assert popular_genre_2 == "Action"
```
Comment (apradoada) [inline tests/test_wave_02.py:78]: I really like this test! This is a great example of an edge case that we don't look for but y'all have decided is important to your code! Well done!

File: tests/test_wave_03.py [hunk 6]
```diff
@@ -54,13 +54,14 @@ def test_friends_unique_movies_not_duplicated():
 
     # Assert
     assert len(friends_unique_movies) == 3
-
-    raise Exception("Test needs to be completed.")
+    assert INTRIGUE_3 in friends_unique_movies
+    assert HORROR_1 in friends_unique_movies
+    assert FANTASY_4 in friends_unique_movies
```
Comment (apradoada) [inline tests/test_wave_03.py:59]: 👍!

File: tests/test_wave_05.py [hunk 7]
```diff
@@ -53,12 +53,16 @@ def test_new_genre_rec_from_empty_friends():
         ]
     }
 
-    raise Exception("Test needs to be completed.")
+    # Act
+    recommendations = get_new_rec_by_genre(sonyas_data)
+
+    # Assert
+    assert len(recommendations) == 0
```
Comment (apradoada) [inline tests/test_wave_05.py:60]: 👍

File: viewing_party/party.py [hunk 8]
```diff
@@ -1,23 +1,147 @@
...
+    movie = {}
 
+    if title and genre and rating:
+        movie["title"] = title
+        movie["genre"] = genre
+        movie["rating"] = rating
+        return movie
+    
+    else:
+        return None
```
Comment (apradoada) [inline viewing_party/party.py:13]: This works well! The one tweak you might want to consider is a common pattern with functions! If we have a situation that would cause the function to stop early, it's best practice to make that a guard clause and include the regular logic of the function outside of the guard clause. This often means checking the negative rather than the positive! For this function, the pattern would be:

```suggestion
    if not title or not genre or not rating:
        return None
        
    movie["title"] = title
    movie["genre"] = genre
    movie["rating"] = rating
    return movie

```

File: viewing_party/party.py [hunk 9]
```diff
@@ -1,23 +1,147 @@
...
 
 def create_movie(title, genre, rating):
-    pass
+    movie = {}
 
+    if title and genre and rating:
+        movie["title"] = title
+        movie["genre"] = genre
+        movie["rating"] = rating
+        return movie
```
Comment (apradoada) [inline viewing_party/party.py:10]: When we see something like this, returning the hard coded dictionary is going to be a more conventional approach:

```suggestion
        return {
             "title": title,
             "genre": genre,
             "rating": rating
        }
```

File: viewing_party/party.py [hunk 10]
```diff
@@ -1,23 +1,147 @@
...
+    
+    else:
+        return None
+
+def add_to_watched(user_data, movie):
+    if movie not in user_data["watched"]:
+        user_data["watched"].append(movie)
```
Comment (apradoada) [inline viewing_party/party.py:17]: I love the check to see if the movie is already there! This is a nice addition that elevates what y'all are already doing!

File: viewing_party/party.py [hunk 11]
```diff
@@ -1,23 +1,147 @@
...
+
+    return user_data
+
+def add_to_watchlist(user_data, movie):
+    if movie not in user_data["watchlist"]:
```
Comment (apradoada) [inline viewing_party/party.py:22]: Nice check here too!

File: viewing_party/party.py [hunk 12]
```diff
@@ -1,23 +1,147 @@
...
+        user_data["watchlist"].append(movie)
+
+    return user_data
+    
+def watch_movie(user_data, title):
+    for movie in user_data["watchlist"]:
+        if movie["title"] == title:
+            add_to_watched(user_data, movie)
+            user_data["watchlist"].remove(movie)
```
Comment (apradoada) [inline viewing_party/party.py:31]: You've made a very smart choice here to use a `for...in` loop rather than a ranged for loop! If we tried to use indexing within this loop, we would run into what is known as a concurrent modification error because we are modifying the list we are looping through. By not removing items by the index, you avoid this particular issue!

File: viewing_party/party.py [hunk 13]
```diff
@@ -1,23 +1,147 @@
...
+        return 0.0
+    
+    sum = 0.0
+
+    for movie in user_data["watched"]:
+        sum += movie["rating"]
+    
+    avg_rating = sum / len(user_data["watched"])
+
+    return avg_rating
```
Comment (apradoada) [inline viewing_party/party.py:50]: 👍

File: viewing_party/party.py [hunk 14]
```diff
@@ -1,23 +1,147 @@
...
+    for movie in user_data["watched"]:
+        genre = movie["genre"]
+
+        if genre not in counts:
+            counts[genre] = 0
+            
+        counts[genre] += 1
+
+        if counts[most_watched_genre] < counts[genre]:
+            most_watched_genre = genre
```
Comment (apradoada) [inline viewing_party/party.py:68]: I understand the instinct to include this particular check within the loop. With that in mind, it can often be a good idea to try and follow the single responsibility principle within our functions as well. We don't necessarily need to make a separate function to handle this portion, but having different parts of our function handle specific actions can really help with the readability of our code. 

By including this particular check within the loop, we technically could potentially end up checking genres against themselves quite a few times. While this doesn't affect our complexity, it's a good idea to try avoiding these extra checks if at all possible! 

We know that the number of genres we have will always be either equal to or less than the number of movies in our list, so we can assume that looping through the genres after the frequency dictionary has been created also won't affect the complexity. 

With both these points in mind, I recommend creating the frequency dictionary first (which can be done with a very short loop) and then iterating through it to find the most_watched_genre. This will increase readability without affecting complexity!

File: viewing_party/party.py [hunk 15]
```diff
@@ -1,23 +1,147 @@
...
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
+def get_unique_watched(user_data):
+    friends_movies = get_friends_movies_list_no_duplicates(user_data["friends"])
+    unique_movies = get_unique_list_1_movies(user_data["watched"], friends_movies)
+
+    return unique_movies
```
Comment (apradoada) [inline viewing_party/party.py:80]: I love the way y'all have used your helper functions here to really streamline these particular functions (This one and the `get_friends_unique_watched`)!

File: viewing_party/party.py [hunk 16]
```diff
@@ -1,23 +1,147 @@
...
+
+def get_friends_movies_list_no_duplicates(friends_list):
+    friends_movies = []
+
+    for friend in friends_list:
+        for movie in friend["watched"]:
+            if movie not in friends_movies:
+                friends_movies.append(movie)
+
+    return friends_movies
```
Comment (apradoada) [inline viewing_party/party.py:96]: I think that overall, getting a collection of all the unique movies for the friends is a great idea. And this function works well. 

One small concern though is that the `not in` step on line 92 does technically make this a triply nested loop which is not ideal. An obvious replacement would be to use sets as they are implemented using Hash maps and allow us to add while avoiding duplicate items in constant time. We do have a slight drawback in the fact that sets do not work with mutable elements (We couldn't add dictionaries to a set for example). 

With this in mind, we could create a set of movie titles that could be used later. 

*Note that this particular approach would require us to modify the helper function and potentially the functions that call this one as well!

File: viewing_party/party.py [hunk 17]
```diff
@@ -1,23 +1,147 @@
...
+    for movie_1 in list_1:
+        is_unique = True
+
+        for movie_2 in list_2:
+            if movie_1 == movie_2:
+                is_unique = False
+                break
+
+        if is_unique:
+            unique_movies.append(movie_1)
```
Comment (apradoada) [inline viewing_party/party.py:110]: Putting my previous suggestion about working with sets instead of lists aside, consider using the `in` method here as opposed to a nested loop! 

```suggestion
    for movie_1 in list_1:
        if movie_1  not in list_2:
            unique_movies.append(movie_1)
            
```

File: viewing_party/party.py [hunk 18]
```diff
@@ -1,23 +1,147 @@
...
 
+def get_available_recs(user_data):
+    recommendations = []
+    friends_unique_movies = get_friends_unique_watched(user_data)
+
+    for movie in friends_unique_movies:
+        if movie["host"] in user_data["subscriptions"]:
+            recommendations.append(movie)
+
+    return recommendations
```
Comment (apradoada) [inline viewing_party/party.py:126]: Perfect!

File: viewing_party/party.py [hunk 19]
```diff
@@ -1,23 +1,147 @@
...
+def get_new_rec_by_genre(user_data):
+    recommendations = []
+    most_watched_genre = get_most_watched_genre(user_data)
+    friends_unique_movies = get_friends_unique_watched(user_data)
+
+    for movie in friends_unique_movies:
+        if movie["genre"] == most_watched_genre:
+            recommendations.append(movie)
+
+    return recommendations
```
Comment (apradoada) [inline viewing_party/party.py:141]: Well done on this function! Great use of the previous functions you have written!

File: viewing_party/party.py [hunk 20]
```diff
@@ -1,23 +1,147 @@
...
+
+def get_rec_from_favorites(user_data):
+    friends_movies = get_friends_movies_list_no_duplicates(user_data["friends"])
+    recommended = get_unique_list_1_movies(user_data["favorites"], friends_movies)
```
Comment (apradoada) [inline viewing_party/party.py:145]: I love this use of a helper function you wrote to help you with wave 3 to help you again here! This is well written and really cuts down on what you need to do for the `get_rec_from_favorites` function!

**Review by apradoada — COMMENTED** [review]
Overall, really well done, y'all! This project will receive a Green! Feel free to reach out if y'all have any questions about your feedback!
